### DISCUSSION TITLE
Post 0.1 Priorities

### DISCUSSION THREAD

**Discussion by ghost** [nvim-neorg/neorg#25] [Ideas]
It looks like 0.1 is getting close to done, at least in number of roadmap items left to complete. It would be nice for potential contributors to get an idea of what order features should/will be prioritized in, and if any orgmode features might be intentionally left out or taken in different design directions. 

Here are a bunch of features that I associate with org mode and its companion packages (some of these are already mentioned in the roadmap).

### Movement/editing shortcuts
Most of these could probably be implemented fairly easily with Tree-Sitter?
- Move cursor to next/previous heading
- Move cursor to next/previous heading at the same level
- Move cursor to parent heading
- Move current heading and content,  up or down 
- Demote or promote current heading (i.e.  change `*** My heading` to `** My heading`)
- Insert new heading
### Finding
I'm including this because I'm not sure how complete 0.1's planned Telescope integration is. 

Also would it be a good idea to move Telescope integration into its own package like org-helm or org-ivy so others could make fzf/fzy/snap integrations if wanted without Telescope being pulled in?
- Fuzzy find heading/tag/name/mark in file/workspace 
### Viewing
Another easy Tree-sitter integration?
- Outline folding
- Outline filtering (sparse trees)
It looks like Neorg's default paradigm is many small files in contrast to Org's a few big files approach, so maybe filtering does not need to be much of a priority
## Note Taking
A lot of org-roam like features built-in would be a nice advantage over org mode. Is an option for database integration like org-roam uses a possibility? 
- Daily Notes
- Inbox/Quick Notes
- Backlinks in sidebar/fuzzy finder
- Fuzzy find existing note and insert link
- Create note, edit it in a pop up window, save it, then insert link to it in current file
- Citation support
- Embedded Latex support
- Footnotes
- External access protocols for saving links or webpages from the browser
- Linting for broken links
## Task Management
- Scheduling Tasks
- Scheduling Recurring Tasks
- Tracking Completion
- Tracking time spent on a task
- Estimating Effort for a task
- Column View 
- Assigning and managing resources
- Enforcing dependencies
- Todo keywords/states/priorities
- Triggers on state changes
- Archiving Tasks
## Tags
I'm not clear on how  heading level org-mode tags fit with the current tag system for Neorg. Would there be a `@heading.meta` tag to put tags/effort/completion information in? Or will Neorg go a different route?
- Subheadings inherit tags from parents
- Shortcuts to select and add tags/priorities/state to headings from a preconfigured list
## Literate Programming
- Tangle
What parameters need to be added to `@code` tags for tangling options?
- Evaluate Code blocks
I know there are some plugins that do this already, but I can't find them at the moment. Will edit when I do
- Lsp in code blocks?
Would be wonderful to get language specific completions
## Import/Export
Seems like big tasks that could maybe wait until later on, but good to keep in mind that any features will need to be translatable to all exports/imports. Org has built-in support for:
>     ascii (ASCII format)
>     beamer (LaTeX Beamer format)
>     html (HTML format)
>     icalendar (iCalendar format)
>     latex (LaTeX format)
>     md (Markdown format)
>     odt (OpenDocument Text format)
>     org (Org format)
>     texinfo (Texinfo format)
>     man (Man page format) 
## UI
 -  Agenda View
 -  Calendar View
 Per #18, the plan might be to have an interface more similar to calcurse than emacs' agenda view
 -  Tables
See discussion [here](https://github.com/vhyrro/neorg/discussions/17)

Where do these features stand on the easy -> hard and nice -> necessary spectrums? I know this is a lot, but I just wanted to get discussions started. Not expecting a complete answer right away.

**Comment by vhyrro** [comment]
Sorry for not responding any faster, I've been a bit too busy with other things, I barely managed to cram in a few commits yesterday lol

### Movement/editing shortcuts
These are fairly easy depending on what you want to do. Simply jumping between different headings should not pose any sort of problem and should be quickly implementable. Jumping to headings on the same level however poses a bit of an issue. You see, our TreeSitter parser isn't perfect, and the syntax trees generated by it are "ok" at best. There's no real understanding of what heading is at the same level (yet). This means that work needs to be done on the TS parser first before we even think about features like that.

### Finding
Don't worry, I'm planning on making the Telescope module a completely separate thing (one that you manually need to pull down). The finder should be able to find everything you specified and also automatically insert links to those locations in the document.

### Viewing
Different views are planned, but not planned for the near future. I don't see a big advantage of implementing that now (it would actually cause a bit of pain since you'd have to do a fair amount of buffer swapping lol).

### Note Taking + Task Management
org-agenda and org-roam-like features are planned to arrive in version 0.2, so you'll have to stick around just a bit more before we get anything magical like that. Yes, this stuff is planned to be implemented straight into the core (no extra downloading of packages necessary).

### Tags
I'm actually thinking about tags and I have a few nice ideas. The initial plan is to have a carryover tag (`$like this`) for headings to give them special attributes. I'm also thinking of different techniques for adding these "tags" to headings (we'll see how we go) :)

### Literate Programming
I'm planning on squeezing in tangling into the 0.1 release (no code block evaluation, just being able to write configs etc. in Neorg). It sounds rather trivial to implement.

LSP in code blocks will be rather difficult hah, I either have to override a bunch of LSP handlers (absolutely a bad idea) or would have to do some buffer + extmark syncing magic. That doesn't mean I'm not gonna attempt it later down the line though.

### Import/Export
We would love to get pandoc support somewhere in the near future for .norg files. It's like a dream thing of mine to do. I understand that file format conversion is one of the most crucial parts of `org-mode` and other organizational tools for many, and I will not overlook it. I think we'll have an in-house .norg to HTML converter simply because speed. The rest will be handled by pandoc (fingers crossed).

### UI
I'm actually currently working on a UI module for Neorg and it's going pretty nicely :D

I don't know whether we should go with an existing API like nui.nvim. Ideally that would cut down a lot of development time but pulling in a bunch of dependencies isn't something a lot of people like.

I hope that at least partially answers the questions you had. If there's anything extra that comes to mind or things you want to know then lemme know :)
